eclair-core/src/main/scala/fr/acinq/eclair/payment/Bolt12Invoice.scala

```diff
@@ -103,65 +80,54 @@ case class PaymentBlindedRoute(route: Sphinx.RouteBlinding.BlindedRoute, payment
-      Some(Paths(paths.map(_.route))),
-      Some(PaymentPathsInfo(paths.map(_.paymentInfo))),
-      offer.issuer.map(Issuer),
-      Some(NodeId(nodeKey.publicKey)),
-      request.quantity_opt.map(Quantity),
-      Some(PayerKey(request.payerKey)),
-      request.payerNote.map(PayerNote),
-      Some(CreatedAt(TimestampSecond.now())),
-      Some(PaymentHash(Crypto.sha256(preimage))),
-      Some(Cltv(minFinalCltvExpiryDelta)),
-      request.payerInfo.map(PayerInfo),
-      request.replaceInvoice.map(ReplaceInvoice),
+    require(request.amount.nonEmpty || request.offer.amount.nonEmpty)
+    val amount = request.amount.orElse(request.offer.amount.map(_ * request.quantity)).get
+    val tlvs: Seq[InvoiceTlv] = removeSignature(request.records).records.toSeq ++ Seq(
+      Some(InvoicePaths(paths.map(_.route))),
+      Some(InvoiceBlindedPay(paths.map(_.paymentInfo))),
+      Some(InvoiceCreatedAt(TimestampSecond.now())),
+      Some(InvoiceRelativeExpiry(300)),
+      Some(InvoicePaymentHash(Crypto.sha256(preimage))),
+      Some(InvoiceAmount(amount)),
+      if (!features.isEmpty) Some(InvoiceFeatures(features.unscoped())) else None,
+      Some(InvoiceNodeId(nodeKey.publicKey)),
     ).flatten
-    val signature = signSchnorr(signatureTag("signature"), rootHash(TlvStream(tlvs), OfferCodecs.invoiceTlvCodec), nodeKey)
+    val signature = signSchnorr(signatureTag, rootHash(TlvStream(tlvs), OfferCodecs.invoiceTlvCodec), nodeKey)
     Bolt12Invoice(TlvStream(tlvs :+ Signature(signature)))
   }
 
   def validate(records: TlvStream[InvoiceTlv]): Either[InvalidTlvPayload, Bolt12Invoice] = {
-    if (records.get[Amount].isEmpty) return Left(MissingRequiredTlv(UInt64(8)))
-    if (records.get[Description].isEmpty) return Left(MissingRequiredTlv(UInt64(10)))
-    if (records.get[Paths].isEmpty) return Left(MissingRequiredTlv(UInt64(16)))
-    if (records.get[PaymentPathsInfo].map(_.paymentInfo.length) != records.get[Paths].map(_.paths.length)) return Left(MissingRequiredTlv(UInt64(18)))
-    if (records.get[NodeId].isEmpty) return Left(MissingRequiredTlv(UInt64(30)))
-    if (records.get[CreatedAt].isEmpty) return Left(MissingRequiredTlv(UInt64(40)))
-    if (records.get[PaymentHash].isEmpty) return Left(MissingRequiredTlv(UInt64(42)))
+    InvoiceRequest.validate(TlvStream[InvoiceRequestTlv](records.records.collect { case tlv: InvoiceRequestTlv => tlv }, records.unknown.filter(_.tag < UInt64(160)))).fold(
+      invalidTlvPayload => return Left(invalidTlvPayload),
+      _ -> ()
+    )
+    if (records.get[InvoiceAmount].isEmpty) return Left(MissingRequiredTlv(UInt64(170)))
+    if (records.get[InvoicePaths].isEmpty) return Left(MissingRequiredTlv(UInt64(160)))
+    if (records.get[InvoiceBlindedPay].map(_.paymentInfo.length) != records.get[InvoicePaths].map(_.paths.length)) return Left(MissingRequiredTlv(UInt64(162)))
+    if (records.get[InvoiceNodeId].isEmpty) return Left(MissingRequiredTlv(UInt64(176)))
```

Comment:
This isn't required anymore.

Reply:
Actually it is in the latest latest version: https://github.com/rustyrussell/lightning-rfc/pull/13
I know it's confusing that rusty didn't put it in the main PR.